kotlin-analysis-core/src/main/kotlin/org/jetbrains/research/ml/kotlinAnalysis/Analyzers.kt

```diff
@@ -0,0 +1,119 @@
+package org.jetbrains.research.ml.kotlinAnalysis
+
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiRecursiveElementVisitor
+
+typealias AnalyzerToStat<P, R> = Map<PsiSecondaryAnalyzer<P, R>, MutableMap<P, R>>
+
+/**
+ * Classes that inherit from this interface implement secondary analyzers that analyze the concrete type of PSI elements.
+ *
+ * @param P the type of PSI elements, that can be analyzed.
+ * @param R the type of the analysis result.
+ */
+interface PsiSecondaryAnalyzer<P : PsiElement, R> {
+    /**
+     * Returns the analysis result of a [PSI element][psiElement].
+     */
+    fun analyze(psiElement: P): R
+}
+
+/**
+ * Classes that inherit from this interface implement secondary analyzers with a cache.
+ * These analyzers analyze the concrete type of PSI elements. The cache is used to memorize the results of the analysis.
+ *
+ * @param P the type of PSI elements, that can be analyzed.
+ * @param R the type of the analysis result.
+ * @param T the type of the cache key.
+ */
+abstract class PsiSecondaryAnalyzerWithCache<P : PsiElement, R, T : Any> : PsiSecondaryAnalyzer<P, R> {
+    abstract val cache: MutableMap<T, R>
+
+    /**
+     * Method to compute key by PSI element. This key would be used for memorization of the analysis result
+     * for the corresponding PSI element.
+     */
+    abstract fun PsiElement.cacheKey(): T
+
+    /**
+     * Returns the analysis result of a [PSI element][psiElement].
+     * If the analysis result is already computed for the corresponding key,
+     * the method returns the value stored in the cache.
+     * Otherwise, it runs [analyzeIfNotCached].
+     */
```

Comment:
Replace into 
```
Analyzes a psi element and returns the analysis result, for example, get a full qualified name for a class or function. If the analysis result is already computed for the corresponding key, the method returns the value stored in the cache. Otherwise, it runs [analyzeIfNotCached].
```
since it is unclear what you mean [PSI element][psiElement]